test/libsolidity/syntaxTests/experimental/inference/polymorphic_type_instantiation_and_operators.sol

```diff
@@ -24,6 +24,8 @@ instantiation str: P2 {}
 instantiation str: P4 {}
 
 instantiation T(A: P1): + {
+    // FIXME: Type comparison with type class function fails because we get
+    // two different variables for A.
     function add(x: T(A), y: T(A)) -> T(A) {}
 }
```

Comment:
This is where I stopped last week. The comparison between function type in the instantiation and the type in the class (after substitution) fails because type `T(A)` has a different type variable in each of these cases. The function from the instantiation has a fixed variable (like all function definitions now) while the type we substitute into the function from the class comes from instantiation header and is generalized.